common/flatpak-transaction.c

```diff
@@ -4062,6 +4092,272 @@ flatpak_transaction_normalize_ops (FlatpakTransaction *self)
+        {
+          g_autofree char *runtime_remote = NULL;
+
+          if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+            {
+              related = find_related_from_deploy (self, runtime_ref, &runtime_remote);
+            }
+          else
+            {
+              FlatpakTransactionOperation *runtime_op = NULL;
+              runtime_op = flatpak_transaction_get_last_op_for_ref (self, runtime_ref);
+              if (runtime_op != NULL)
+                {
+                  runtime_remote = g_strdup (runtime_op->remote);
+                  if (!op_get_related (self, runtime_op, &related, error))
+                    return FALSE;
+                }
+              else
+                related = find_related_from_deploy (self, runtime_ref, &runtime_remote);
+            }
+
+          if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+            {
+              g_hash_table_replace (maybe_unused_runtimes, g_strdup (runtime_ref), g_strdup (runtime_remote));
+              g_ptr_array_add (run_after_ops, op);
+            }
+          else
+            g_hash_table_add (newly_used_runtimes, g_strdup (runtime_ref));
+
+          if (!priv->disable_related)
+            {
+              for (i = 0; related && i < related->len; i++)
+                {
+                  FlatpakRelated *rel = g_ptr_array_index (related, i);
+                  if (!rel->delete)
+                    continue;
+
+                  if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+                    g_hash_table_replace (maybe_unused_runtimes, g_strdup (rel->ref), g_strdup (runtime_remote));
+                  else
+                    g_hash_table_add (newly_used_runtimes, g_strdup (rel->ref));
+                }
+            }
+        }
+
+      /* If the op is changing to a different runtime, the previous one might
+       * now be unused */
+      if (op_type == FLATPAK_TRANSACTION_OPERATION_UPDATE &&
+          dir_ref_is_installed (priv->dir, op->ref, NULL, &deploy_data))
+        {
+          g_autofree char *full_previous_runtime = NULL;
+          g_autofree char *prev_runtime_remote = NULL;
+
+          const char *previous_runtime = flatpak_deploy_data_get_runtime (deploy_data);
+
+          if (previous_runtime == NULL || *previous_runtime == '\0')
+            continue;
+
+          full_previous_runtime = g_strconcat ("runtime/", previous_runtime, NULL);
+          if (g_strcmp0 (full_previous_runtime, runtime_ref) == 0)
+            continue;
+
+          g_clear_pointer (&related, g_ptr_array_unref);
+          related = find_related_from_deploy (self, full_previous_runtime, &prev_runtime_remote);
+
+          g_hash_table_replace (maybe_unused_runtimes,
+                                g_strdup (full_previous_runtime),
+                                g_strdup (prev_runtime_remote));
+
+          g_ptr_array_add (to_be_excluded, op->ref);
+
+          g_ptr_array_add (run_after_ops, op);
```

Comment:
We could have added the op to the list already above.

Reply:
Not sure what you mean

Reply:
Oh, i see. I was wrong, we only add the op above for uninstall ops, and this only for update ops, so it won't get added twice